### PR TITLE
Image Block: Enable image block to be selected correctly when clicked

### LINKED ISSUE #56017: [WP 6.4.1] images can't be corrected selected as Gutenberg block after linking to media
### Description

Images can't be corrected selected as Gutenberg block after linking to media. This appears after wp 6.4. 

### Step-by-step reproduction instructions

1. Create an image in Gutenberg editor
2. Click the image: currently the image can be selected as a Gutenberg block, and blue selection border appears. 
3. Link the image to its media file
4. Click the image again: currently the image cannot be selected as a Gutenberg block, and blue selection border disappears (need to press Up Arrow key to restore it). 

### Screenshots, screen recording, code snippet

_No response_

### Environment info

New installed WP 6.4.1 in docker, without any other plugins or themes
Chrome
macOS Sonoma

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by t-hamano**
Follow-up on #55470
Fixes: #56017

## What?

This PR allows you to select image blocks "correctly".

This will display a blue outline when a block is selected, and inserting a block with the Enter key and deleting a block with the Delete key will work correctly.

## Why?

With #55470, the image block is now wrapped in an A tag when it's linked. As a result, even if you click on the block, it no longer appears to be selected correctly.

## How?

To find a way to solve this, I looked into blocks where the block itself is an A element, or where there is an A element inside the block. As a result, I noticed that there are many approaches that add `pointer-events:none` to the A element.

I can't explain it exactly, but I found that block selection works correctly with this CSS.

**Update:** I found that if I simply added `pointer-events: none` to the `a` element, the image could no longer be resized. This is because the `ResizableBox` component is wrapped in an `a` element. To solve this I moved the `a` element inside the `ResisableBox`. Check [this comment](https://github.com/WordPress/gutenberg/pull/56043#issuecomment-1810356718) for details.

## Testing Instructions

- Insert an Image block.
- Set a link to the Image block.
- Remove focus from the Image block.
- "Click" on the image block (Note: If you select the image block with the up and down keys on your keyboard, the block will be selected correctly, so click on the block with your mouse).
- A blue outline will appear to indicate that the block is selected.
- When you press Enter, an empty paragraph block will be generated next to the image block.
- You can delete the image block by pressing the Delete key.

## Screenshots or screencast

https://github.com/WordPress/gutenberg/assets/54422211/8a7a1d8d-2c4c-4f46-9c79-1109b05adfec

### CODE REVIEW

**Comment by t-hamano** [comment]
@alexstine I would appreciate it if you could let me know if applying `pointer-events:none` to the `a` element will cause accessibility issues 🙏

**Comment by alexstine** [comment]
@t-hamano I am not sure about this but I would think so.

CC: @joedolson @afercia @jeryj.

**Comment by jeryj** [comment]
The `pointer-events: none;` may be part of the solution, but it may not be necessary. Using the arrow keys to navigate the blocks shows the issue as the `<a>` is considered a selectable block, when it shouldn't be. 

### Reproduction
- Have Paragraph Block, Image Block with link, Paragraph Block
- Starting on First paragraph block, arrow down.
- Image block will be selected
- Arrow Down again
- Bug: Image block is still selected, but on the `<a>`. No block border is visibile.
- Arrow Down
- Second Paragraph block is selected

I haven't dug into the code, but my guess is that it has to do with the `<a>` within the image block being considered a selectable element. If it's not a selectable element within the flow of the blocks, then clicking on the `<a>` shouldn't do anything.

Could we use `role=link` on the wrapping element and make it a generic `<div>`? Or would that be worse?

**Comment by carolinan** [comment]
The markup in the editor and the font should match as close as possible, otherwise it becomes more difficult to style.

**Comment by alexstine** [comment]
@jeryj In that case, I think writing-flow/use-arrow-nav hook needs to be explored. I think it matches `<a>` as a focusable element. Just not sure how to update the logic without breaking accessibility. It might also be related to `block-list/use-focus-first-element` hook. Using `role="link"` would probably be okay but maybe not for visuals.

**Comment by t-hamano** [comment]
The problem with focusing on the `a` element with the arrow keys is probably because it is considered a focusable element, as @alexstine  said.

I was able to solve this problem by adding `tabIndex="-1"` to the `a` element of the image block, but are there any concerns about this?

The video below shows pressing the up and down arrow keys repeatedly on an editor containing an Image, Site Logo, and Paragraph block. You can see that the focus no longer moves to the `a` element of the Image block. On the other hand, in the Site Logo block where the `a` element is included in the block, you can see that the focus has shifted to the `a` element.

https://github.com/WordPress/gutenberg/assets/54422211/3132bb11-8c16-4453-97f3-e4f7ec374785

File: packages/block-library/src/image/editor.scss
Comment (ramonjd) [inline packages/block-library/src/image/editor.scss]:
```suggestion
		// When the Image block is linked,
		// it's wrapped with a disabled <a /> tag.
		// Restore cursor style so it doesn't appear 'clickable'
		// and remove pointer events. Safari needs the display property.
		pointer-events: none;
		cursor: default;
		display: inline-block;
```

Comment (ramonjd) [reply packages/block-library/src/image/editor.scss]: Maybe we could group the styles so folks know that they belong to the same "change".

What do you think?

Also pointer events none doesn't work in Safari for me with the display property.

File: packages/block-library/src/image/image.js
Comment (ramonjd) [inline packages/block-library/src/image/image.js]: >  was able to solve this problem by adding tabIndex="-1" to the a element of the image block, but are there any concerns about this?

This fixes the issue that @jeryj [reported](https://github.com/WordPress/gutenberg/pull/56043#issuecomment-1806337980) for me.

Seems to be acceptable according to the [spec](https://www.w3.org/TR/2011/WD-html5-20110525/editing.html#sequential-focus-navigation-and-the-tabindex-attribute), but I'll defer to other folks.

File: packages/block-library/src/image/image.js
Comment (ramonjd) [inline packages/block-library/src/image/image.js]: 👍🏻

**Review by ramonjd — COMMENTED** [review]
Thank you @t-hamano for this PR!

> Using role="link" would probably be okay but maybe not for visuals.

Unfortunately, in my testing, changing the tag and adding the role attribute doesn't trigger the styles as the markup is different as @carolinan [notes](https://github.com/WordPress/gutenberg/pull/56043#issuecomment-1806451309).

> The pointer-events: none; may be part of the solution, but it may not be necessary.

In my testing, I couldn't find an alternative to this. Removing the pointer events ensure that the click even is triggered on the parent element. 

We only need to cater for Safari (surprise, surprise) by adding `display: inline-block`. 

I've tested in FF, Chrome and Edge as well and it's working with the above change.

Here's what I'm seeing:

✅ Click bug is fixed (with `display: inline-block` for Safari)
✅ Arrow up and down selects previous and subsequent blocks immediately

Would it be helpful to add `role="presentation"` to the link to remove the `<a />` semantics or am I completely overthinking it? 🤣 

I'm happy to approve this if other folks are fine with the changes.

File: packages/block-library/src/image/image.js
Comment (alexstine) [reply packages/block-library/src/image/image.js]: @t-hamano No, you can't add `tabindex="-1"` like this. Doing that prevents the link from gaining focus with arrow key navigation. You need to be able to focus if you are going to hear things like the URL, ALT text, filename, etc.

Comment (ramonjd) [reply packages/block-library/src/image/image.js]: In the context of the editor, the link is a very recent addition whose only function is to ensure the editor styles match the frontend because some global styles target A-tags. The A-tag doesn't actually need an href value to achieve this.

Would it be acceptable to remove it from the accessibility tree or mark it as presentation only or something? 

Were the ALT text and filename not available from the IMG element previously? 

If getting these attributes is preferable and, bear with me as this might be wild 😄, could we add a data-href to the surrounding figure, which is the selectable "block" element?

Comment (alexstine) [reply packages/block-library/src/image/image.js]: @ramonjd You can't add `role="presentation"` to the link or else you will also destroy the semantics of the `` tag wrapped inside. I never actually knew you could link an image block so not sure how it worked before. Pretty sure if there is no link, you could not access that information because `` by default cannot take focus. That is a bug that needs to be fixed somehow.

Comment (t-hamano) [reply packages/block-library/src/image/image.js]: It's a very difficult problem, but for this PR, it might be better to focus on the problem of not being able to click the linked Image block correctly. That is, don't add code like `tabIndex=-1` that prevents the `a` element in the block from being focused.

It is not only the Image block that the keyboard focuses on the `a` element within the block. For example, a similar problem occurs with the Site Logo block with the home link set.

I think we can discuss more broadly in another issue, such as what kind of keyboard interaction is expected in the editor and what should be communicated to screen readers when the block contains an `a` element or an `img` element. .

Comment (alexstine) [reply packages/block-library/src/image/image.js]: Agreed. If you use your arrow keys to select the block and the `<a>` becomes focussed, you can always focus the block wrapper with left arrow. This is how all blocks with wrappers work with `tabindex="0"` set on the wrapper.

Comment (ramonjd) [reply packages/block-library/src/image/image.js]: Ah okay, I appreciate the explanation @alexstine Thank you! 👍🏻 

> I think we can discuss more broadly in another issue, such as what kind of keyboard interaction is expected in the editor and what should be communicated to screen readers when the block contains an a element or an img element. .

Sounds like a good approach.

**Comment by t-hamano** [comment]
I noticed that if I applied `pointer-events: none;` to the `a` element, the resize handle inside would no longer work 😅 Let's try moving this `a` element into a `ResizableBox`.

https://github.com/WordPress/gutenberg/assets/54422211/a795edc8-4e7c-4c78-9b53-3281a0f243a8

**Comment by t-hamano** [comment]
> I noticed that if I applied `pointer-events: none;` to the `a` element, the resize handle inside would no longer work 😅 Let's try moving this `a` element into a `ResizableBox`.

I tried to solve this problem by wrapping the `img` element directly in an `a` tag if the image has an href. In other words, the way it is rendered will change as follows.

Before:

```jsx
<a style={{pointerEvents: 'none'}}>
  <ResisableBox>
    <img />
  </ResisableBox>
</a>
```

After:

```jsx
<ResisableBox>
  <a style={{pointerEvents: 'none'}}>
    <img />
  </a>
</ResisableBox>
```

Other changes are as follows:

- The `tabIndex="-1"` of the `a` element added in this PR has been removed again. The reason is mentioned in [this comment](https://github.com/WordPress/gutenberg/pull/56043#discussion_r1390601343).
- Changed styles such as `pointer-events: none;` from the stylesheet to inline styles. This is because the `a` element is not necessarily directly below the block, and CSS selectors like `.wp-block-image > a` cannot be used.

**Comment by t-hamano** [comment]
It seems to work fine in my testing. In all scenarios, the Image block is clickable, and the border color also seems to inherit the link color.

In the screenshot below, I'm testing various scenarios with the Twenty Twenty-Three theme, which has a green link color.

https://github.com/WordPress/gutenberg/assets/54422211/a892becc-82e7-4aad-b34d-bf76343925d0

File: packages/block-library/src/image/image.js
Comment (alexstine) [inline packages/block-library/src/image/image.js:102]: @t-hamano Do you think this could be confusing for low vision users to get a normal mouse pointer when the link is not clickable?

**Review by alexstine — APPROVED** [review]
@t-hamano One question but otherwise this is testing well with screen readers.

File: packages/block-library/src/image/image.js
Comment (ramonjd) [reply packages/block-library/src/image/image.js:102]: On the face of it, `default` seems appropriate to me given that the link is disabled. (?)

`cursor: 'pointer'` is the "hand", or default browser cursor, for linked elements.

**Comment by ramonjd** [comment]
I can confirm that there's no regression to the fix for https://github.com/WordPress/gutenberg/issues/55336 as well.

Just to confirm I can still reproduce the arrow selection scenario described [above](https://github.com/WordPress/gutenberg/pull/56043#issuecomment-1806337980), but that's for a [follow up](https://github.com/WordPress/gutenberg/pull/56043#discussion_r1391247979), right?

Thanks @t-hamano and @alexstine for sticking with this 🍺

File: packages/block-library/src/image/image.js
Comment (t-hamano) [reply packages/block-library/src/image/image.js:102]: I think that since this element cannot actually be clicked, it is okay to leave the cursor as the default.

By the way, #56123 has been submitted to add a pressed state to the link button in the block toolbar to make it easier to tell if the image is _linked_.

**Comment by t-hamano** [comment]
> Just to confirm I can still reproduce the arrow selection scenario described [above](https://github.com/WordPress/gutenberg/pull/56043#issuecomment-1806337980), but that's for a [follow up](https://github.com/WordPress/gutenberg/pull/56043#discussion_r1391247979), right?

That's right. The behavior of being able to select the `a` element with the arrow keys is not the first time it has occurred in this PR, but may occur in all blocks that include the `a` element. So what scenario is ideal or not would be discussed in a separate issue/PR.

**Comment by t-hamano** [comment]
I checked again after merging the latest trunk and discovered another problem with this PR. When I apply a link to the image, the height of the image changes and it extends beyond the `ResisableBox`.

https://github.com/WordPress/gutenberg/assets/54422211/5f8d3d60-6ede-4e99-8242-c80ccfae5292

I expected that `display:inline-block` was having an effect and added `vertical-align: top`, but it had no effect. `display:block` also had no effect. In the end, I was able to solve the problem by adding `display: inline`.

@ramonjd In [this comment](https://github.com/WordPress/gutenberg/pull/56043#pullrequestreview-1726362122) of yours, you say that the `display` property is needed to solve the click bug in ~Firefox~ Safari.

Could you please confirm whether `dipslay:inline` also solves this problem? Because I'm a Windows user and Safari browser doesn't exist 😅

**Comment by t-hamano** [comment]
Another approach might be to focus the block when the `a` element is clicked, instead of making it impossible to click on it. I'll try to find a more ideal approach.

**Comment by ramonjd** [comment]
> @ramonjd In https://github.com/WordPress/gutenberg/pull/56043#pullrequestreview-1726362122 of yours, you say that the display property is needed to solve the click bug in Firefox.

I think you mean Safari? No problem, I'll test today and get back to you!

**Comment by ramonjd** [comment]
>  I was able to solve the problem by adding display: inline.

It looks good to me

Here's what I'm seeing in Safari:

With `display: inline`

https://github.com/WordPress/gutenberg/assets/6458278/45bda350-4e2f-4dfb-be4b-b9597b89ba4f



With `display: inline-block`
Screenshot 2023-11-17 at 7 07 49 am

**Comment by t-hamano** [comment]
Thanks for the testing, @ramonjd!

> I think you mean Safari?

Yes, that's right 😅

> >  I was able to solve the problem by adding display: inline.
> It looks good to me

I was relieved that it seemed to work well. I believe that all problems have been resolved, so I would like to merge this PR.

Thank you for your cooperation!

**Comment by t-hamano** [comment]
Note: As stated in [this comment](https://github.com/WordPress/gutenberg/issues/56017#issuecomment-1805933527), this PR fixes a bug that first appeared in WP6.4. I'm giving it a backport label because I think this should be fixed in a minor release.

**Comment by t-hamano** [comment]
@mikachan Is it possible to backport this PR to WP6.4.2?

**Comment by mikachan** [comment]
> @mikachan Is it possible to backport this PR to WP6.4.2?

Yes that should be fine, as 6.4.2 is still being worked on. I'll cherry-pick this PR for inclusion in a package update, ready for 6.4.2.

**Comment by mikachan** [comment]
I just cherry-picked this PR to the 6-4-2-cherry-picks branch to get it included in the next release: d5de4af46e

**Pull request merged**
